### ISSUE TITLE
Coq 8.12 "-I" option inserts ML path after user-contrib

### ISSUE THREAD

**Issue by akr** [coq/coq#12771]
#### Description of the problem

I found Coq 8.12's "-I" option inserts the directory after user-contrib directories.

```
% bin/coqtop -I /tmp
Welcome to Coq 8.12.0 (July 2020)

Coq < Print ML Path.
ML Load Path:
  /home/src/coq/812/lib/coq/user-contrib
  /home/src/coq/812/lib/coq/user-contrib/Ltac2
  /home/src/coq/812
  /tmp
  /home/src/coq/812/lib/coq/plugins/rtauto
  /home/src/coq/812/lib/coq/plugins/syntax
  /home/src/coq/812/lib/coq/plugins/cc
  /home/src/coq/812/lib/coq/plugins/funind
  /home/src/coq/812/lib/coq/plugins/ssrsearch
  /home/src/coq/812/lib/coq/plugins/firstorder
  /home/src/coq/812/lib/coq/plugins/derive
  /home/src/coq/812/lib/coq/plugins/omega
  /home/src/coq/812/lib/coq/plugins/setoid_ring
  /home/src/coq/812/lib/coq/plugins/micromega
  /home/src/coq/812/lib/coq/plugins/nsatz
  /home/src/coq/812/lib/coq/plugins/btauto
  /home/src/coq/812/lib/coq/plugins/ltac
  /home/src/coq/812/lib/coq/plugins/ssrmatching
  /home/src/coq/812/lib/coq/plugins/extraction
  /home/src/coq/812/lib/coq/plugins/ssr
```

Coq 8.11.1's "-I" option inserts the directory at the beginning.

```
% bin/coqtop -I /tmp
Welcome to Coq 8.11.1 (July 2020)

Coq < Print ML Path.
ML Load Path:
  /tmp
  /home/src/coq/8111/lib/coq/user-contrib
  /home/src/coq/8111/lib/coq/user-contrib/Ltac2
  /home/src/coq/8111/lib/coq/plugins
  /home/src/coq/8111/lib/coq/plugins/rtauto
  /home/src/coq/8111/lib/coq/plugins/syntax
  /home/src/coq/8111/lib/coq/plugins/cc
  /home/src/coq/8111/lib/coq/plugins/funind
  /home/src/coq/8111/lib/coq/plugins/firstorder
  /home/src/coq/8111/lib/coq/plugins/derive
  /home/src/coq/8111/lib/coq/plugins/omega
  /home/src/coq/8111/lib/coq/plugins/setoid_ring
  /home/src/coq/8111/lib/coq/plugins/micromega
  /home/src/coq/8111/lib/coq/plugins/nsatz
  /home/src/coq/8111/lib/coq/plugins/btauto
  /home/src/coq/8111/lib/coq/plugins/ltac
  /home/src/coq/8111/lib/coq/plugins/ssrmatching
  /home/src/coq/8111/lib/coq/plugins/extraction
  /home/src/coq/8111/lib/coq/plugins/ssr
  /home/src/coq/8111
```

I found this difference at my plugin development.
(Gallna-to-C code generator: https://github.com/akr/codegen )

The difference of "-I" behavior makes my plugin development difficult
because plugin testing require plugin instllation.
(after once I install the plugin to user-contrib directory.)

With Coq 8.11, I can test non-installed development version of my plugin
using "-I" option, even when user-contrib directory has an old version of my plugin.

But with Coq 8.12, I cannot do it because Coq find the plugin file, codegen_plugin.cmxs,
in user-contrib directory instead of the directory I specified with "-I" option.

Is this change intentional?

Note that I found a sentence, "Coq's -I option becomes closer to OCaml's semantics." in
https://github.com/coq/coq/pull/11618 .
But I couldn't check OCaml semantics because I couldn't find a way to show OCaml search path in ocamltop.

#### Coq Version

```
% bin/coqtop -v     
The Coq Proof Assistant, version 8.12.0 (July 2020)
compiled on Jul 29 2020 11:34:32 with OCaml 4.10.0
% lsb_release -a
No LSB modules are available.
Distributor ID:	Debian
Description:	Debian GNU/Linux 10 (buster)
Release:	10
Codename:	buster
% uname -mrsv
Linux 4.19.0-9-amd64 #1 SMP Debian 4.19.118-2+deb10u1 (2020-06-07) x86_64
```

**Comment by Zimmi48** [comment]
From the fact that Emilio milestoned and self-assigned this issue, I infer that this is indeed an unintentional change. However, as general advice for plugin development and testing, I'd like to encourage you to try using Dune. This provides, in particular, build compositionality, which means that if you drop a copy of Coq's source code into your plugin folder, you can build the two together and thus test them without installing or being impeded by the previous installations of Coq and Coq packages.

**Comment by akr** [comment]
> From the fact that Emilio milestoned and self-assigned this issue, I infer that this is indeed an unintentional change. However, as general advice for plugin development and testing, I'd like to encourage you to try using Dune. This provides, in particular, build compositionality, which means that if you drop a copy of Coq's source code into your plugin folder, you can build the two together and thus test them without installing or being impeded by the previous installations of Coq and Coq packages.

Thank you.  I'll try it.

**Comment by ejgallego** [comment]
Indeed this change was not intended, it was a side-effect of some code cleanup, thanks for the detailed report.

I intend to amend this in 8.12.1, PR incoming.

**Comment by ejgallego** [comment]
This is unfortunately hard to fix in a minor 8.12 release as it requires a change of API :(

#11618 separated the ML paths into the ones that come with the vo files and the ones that come alone; maybe we could collect the ML paths from the VO and add them in the right order, however that's gonna be a bit invasive.
